File: jyd_plain_json_probe/apps/processor/frontend/new/device-authorization.js
(() => {
  'use strict';
  const el = id => document.getElementById(id);
  let busy = false;
  let state = 'LOADING';
  let timer = null;
  const resuming = new Set();
  const descriptions = {
    UNREGISTERED: ['尚未申请设备授权', '为这台处理机提交一次申请，管理员批准后即可使用。', 'warning'],
    PENDING: ['等待管理员批准', '申请已经提交，无需再次申请或重新安装。批准后会自动继续校验。', 'warning'],
    ACTIVE: ['设备已授权', '原设备身份已确认。正常覆盖更新、移动目录和重新登录无需重复激活。', 'ok'],
    OFFLINE_GRACE: ['暂时无法联网校验', '仅能在已有短期凭据的有效范围内使用获准功能；不会重新开始宽限期。', 'warning'],
    AUTH_REFRESH_REQUIRED: ['需要重新校验授权', '这是凭据刷新，不是重新激活。请保持联网后重新校验。', 'warning'],
    REJECTED: ['设备申请未获批准', '请联系管理员核对申请，不要重复申请。', 'error'],
    SUSPENDED: ['设备授权已暂停', '请联系管理员恢复授权；重新安装不会解除暂停。', 'error'],
    REVOKED: ['设备授权已撤销', '请联系管理员办理授权或换机，原授权不会因重新安装恢复。', 'error'],
    EXPIRED: ['设备授权已到期', '请联系管理员确认授权期限。', 'warning'],
    KEY_UNAVAILABLE: ['无法读取本机设备密钥', '请先检查系统权限或 TPM 状态；程序不会自动删除或替换原密钥。', 'error'],
    KEY_INITIALIZING: ['正在处理 Windows 权限确认', '请在实际运行工作台的电脑上完成确认。程序保留当前初始化过程，不会重复启动或替换原密钥；完成后重新校验。', 'warning'],
    LOGIN_REQUIRED: ['请先登录数字人账号', '登录成功后会复用这台处理机已有的批准关系，不需要重新申请。', 'warning'],
    CLIENT_UPGRADE_REQUIRED: ['请更新工作台程序', '更新程序后保留原设备授权，无需管理员再次批准。', 'warning']
  };
  async function api(action = '') {
    const options = {credentials:'same-origin', cache:'no-store', headers:{}};
    if (action) {
      options.method = 'POST';
      options.headers['X-Device-Authorization-Action'] = '1';
      if (action === '/apply' || action === '/apply-software') {
        options.headers['Content-Type'] = 'application/json';
        const application = {label:el('label').value.trim(), confirm_initialize:true};
        if (action === '/apply-software') application.confirm_software = true;
        options.body = JSON.stringify(application);
      }
      if (action === '/repair-key-access') {
        options.headers['Content-Type'] = 'application/json';
        options.body = JSON.stringify({confirm_repair:true});
      }
    }
    const response = await fetch('/api/new/device-authorization' + action, options);
    const body = await response.json().catch(() => ({}));
    if (response.status === 401 && (!body.code || body.code === 'LOGIN_REQUIRED')) body.state = 'LOGIN_REQUIRED';
    if (!response.ok && !body.state) body.state = 'AUTH_REFRESH_REQUIRED';
    if (!body.state) throw new Error('处理机没有返回有效的授权状态，请更新兼容版本的程序。');
    return body;
  }
  function render(data) {
    state = data.state;
    let info = descriptions[state] || ['授权状态暂不可用', '请重新校验或联系管理员检查服务配置。', 'warning'];
    if (data.code === 'DEVICE_TRUST_NOT_CONFIGURED') info = ['此程序尚未配置正式授权', '请使用包含正式验证公钥的工作台版本。重复申请设备不会解决程序配置问题。', 'warning'];
    if (data.code === 'DEVICE_SOFTWARE_NOT_ALLOWED') info = ['尚未允许软件兼容保护', '请先由网站管理员为此账号允许软件保护，再重新校验并明确申请；本次没有创建或替换设备密钥。', 'warning'];
    el('status-title').textContent = info[0];
    el('status-description').textContent = info[1];
    el('status-dot').dataset.tone = info[2];
    el('application').hidden = state !== 'UNREGISTERED';
    el('key-repair').hidden = (data.code || data.error_code) !== 'KEY_ACCESS_DENIED';
    el('login').hidden = state !== 'LOGIN_REQUIRED';
    el('continue').hidden = state !== 'ACTIVE';
    el('device-id').textContent = data.device_id || (data.thumbprint ? data.thumbprint.slice(0,16) + '…' : '尚未读取');
    el('device-id').title = data.thumbprint || '';
    el('account').textContent = data.user_id ? '账号 ID：' + data.user_id : '尚未读取账号';
    el('protection').textContent = data.protection_report === 'tpm' ? 'TPM（本机报告，未做服务器硬件证明）' : data.protection_report === 'software' ? '软件兼容保护（较弱）' : '—';
    el('expires').textContent = Number.isFinite(data.exp) ? new Date(data.exp * 1000).toLocaleString() : '—';
    el('error').hidden = !data.detail;
    el('error').textContent = typeof data.detail === 'string' ? data.detail : '';
  }
  async function loadWaitingJobs() {
    const panel = el('waiting-jobs');
    if (state === 'LOGIN_REQUIRED') {panel.hidden = true; el('waiting-list').replaceChildren(); return;}
    try {
      const response = await fetch('/api/new/device-authorization/waiting-jobs', {credentials:'same-origin',cache:'no-store'});
      if (!response.ok) {panel.hidden = true; el('waiting-list').replaceChildren(); return;}
      const result = await response.json();
      const jobs = Array.isArray(result.jobs) ? result.jobs : [];
      panel.hidden = jobs.length === 0;
      el('waiting-list').replaceChildren();
      for (const job of jobs) {
        if (typeof job.job_id !== 'string' || !/^[A-Za-z0-9_-]{1,100}$/.test(job.job_id)) continue;
        const row = document.createElement('div'); row.className = 'actions';
        const label = document.createElement('p');
        label.textContent = '任务 ' + job.job_id.slice(0,12) + ' · ' + (job.created_at || '等待继续');
        const button = document.createElement('button'); button.type = 'button'; button.textContent = '继续此任务';
        button.disabled = resuming.has(job.job_id);
        button.addEventListener('click', async () => {
          if (resuming.has(job.job_id)) return;
          resuming.add(job.job_id); button.disabled = true; el('waiting-error').hidden = true;
          try {
            const reply = await fetch('/api/jobs/' + encodeURIComponent(job.job_id) + '/resume-authorization', {
              method:'POST', credentials:'same-origin', headers:{'X-Device-Authorization-Action':'1'}
            });
            const value = await reply.json().catch(() => ({}));
            if (!reply.ok) throw new Error(typeof value.detail === 'string' ? value.detail : '暂时无法继续，请先校验设备授权。');
            await loadWaitingJobs();
          } catch (error) {el('waiting-error').textContent = error.message; el('waiting-error').hidden = false;}
          finally {resuming.delete(job.job_id); button.disabled = false;}
        });
        row.append(label, button); el('waiting-list').append(row);
      }
    } catch (_) {panel.hidden = true; el('waiting-list').replaceChildren();}
  }
  function schedule() {
    clearTimeout(timer);
    if (['PENDING','ACTIVE','AUTH_REFRESH_REQUIRED','OFFLINE_GRACE','KEY_INITIALIZING'].includes(state)) {
      timer = setTimeout(() => {if (!document.hidden) run(); else schedule();}, state === 'ACTIVE' ? 300000 : 15000);
    }
  }
  async function run(action = '') {
    if (busy) return;
    busy = true;
    el('refresh').disabled = true;
    el('apply').disabled = true;
    el('apply-software').disabled = true;
    el('repair').disabled = true;
    try {
      let data = await api(action);
      // Reading a server approval is informational; refresh obtains signed
      // credentials. This never creates a key or makes another application.
      if (!action && ['ACTIVE','AUTH_REFRESH_REQUIRED'].includes(data.state) && data.thumbprint && !data.code && !data.error_code) data = await api('/refresh');
      render(data);
      await loadWaitingJobs();
    } catch (error) {
      el('error').hidden = false;
      el('error').textContent = error.message || '网络暂时不可用，请稍后重试。';
    } finally {
      busy = false;
      if (action === '/repair-key-access') el('repair-confirm').checked = false;
      if (action === '/apply-software') el('software-confirm').checked = false;
      el('refresh').disabled = false;
      el('apply').disabled = !el('confirm').checked;
      el('apply-software').disabled = !el('confirm').checked || !el('software-confirm').checked;
      el('repair').disabled = !el('repair-confirm').checked;
      schedule();
    }
  }
  function updateApplicationButtons() {
    el('apply').disabled = busy || !el('confirm').checked;
    el('apply-software').disabled = busy || !el('confirm').checked || !el('software-confirm').checked;
  }
  el('confirm').addEventListener('change', updateApplicationButtons);
  el('software-confirm').addEventListener('change', updateApplicationButtons);
  el('apply-software').addEventListener('click', () => {
    if (state === 'UNREGISTERED' && el('confirm').checked && el('software-confirm').checked && !busy) run('/apply-software');
  });
  el('apply-form').addEventListener('submit', event => {
    event.preventDefault();
    if (state === 'UNREGISTERED' && el('confirm').checked && !busy) run('/apply');
  });
  el('refresh').addEventListener('click', () => run('/refresh'));
  el('repair-confirm').addEventListener('change', () => {el('repair').disabled = busy || !el('repair-confirm').checked;});
  el('repair-form').addEventListener('submit', event => {
    event.preventDefault();
    if (!el('key-repair').hidden && el('repair-confirm').checked && !busy) run('/repair-key-access');
  });
  document.addEventListener('visibilitychange', () => {if (!document.hidden && !busy) run();});
  window.addEventListener('pagehide', () => clearTimeout(timer));
  run();
})();

File: jyd_plain_json_probe/src/jyd_probe/device_authorization.py
"""Per-account device sessions: refresh credentials, never recreate identity.

No startup key creation, provider calls, business retry, or UI signing endpoint.
The session is consumed by server-side admission, not trusted browser booleans.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
import re
import tempfile
import threading
import time
from urllib.error import HTTPError, URLError
from urllib.request import HTTPRedirectHandler, Request, build_opener

from .device_auth_protocol import (
    LEASE_TYPE,
    DeviceAuthorizationError,
    TrustedIssuer,
    VerifiedCredentials,
    assert_key_available,
    canonical_json,
    make_proof,
    sha256_b64,
    strict_json,
)
from .device_identity_windows import DeviceIdentityError, WindowsDeviceIdentity

BASE = "/api/workbench/device-auth"
MAX_RESPONSE_BYTES = 65536


class _NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        # A redirect may change htu or leak the account token/proof to another host.
        return None


class DeviceAuthTransport:
    def __init__(
        self, trust: TrustedIssuer, *, timeout_seconds: float = 4, opener=None
    ):
        self.trust = trust
        self.timeout_seconds = max(1, min(30, float(timeout_seconds)))
        self._opener = opener if opener is not None else build_opener(_NoRedirect())

    def request(
        self, *, method: str, path: str, headers: dict, payload: dict | None = None
    ) -> dict:
        if path not in {
            BASE + "/" + suffix
            for suffix in (
                "challenge",
                "register",
                "status",
                "exchange",
                "refresh",
                "local-policy",
                "software-initialization-permit",
                "agent-permit",
            )
        }:
            raise DeviceAuthorizationError(
                "UNTRUSTED_DEVICE_AUTH_TARGET", "设备授权接口地址无效"
            )
        uri = self.trust.request_uri(path)
        request = Request(
            uri,
            method=method,
            data=None if payload is None else canonical_json(payload).encode("ascii"),
            headers={
                "Accept": "application/json",
                "Content-Type": "application/json",
                **headers,
            },
        )
        try:
            with self._opener.open(request, timeout=self.timeout_seconds) as response:
                if response.geturl() != uri:
                    raise DeviceAuthorizationError(
                        "DEVICE_AUTH_REDIRECT_REJECTED", "设备授权服务不允许重定向"
                    )
                raw = response.read(MAX_RESPONSE_BYTES + 1)
                status = int(response.status)
        except HTTPError as exc:
            try:
                raw = exc.read(MAX_RESPONSE_BYTES + 1)
            finally:
                exc.close()
            self._raise_response(int(exc.code), raw)
        except (URLError, OSError, TimeoutError) as exc:
            raise DeviceAuthorizationError(
                "DEVICE_AUTH_UNREACHABLE",
                "暂时无法连接设备授权服务",
                status_code=503,
                transient=True,
            ) from exc
        if not 200 <= status < 300:
            self._raise_response(status, raw)
        try:
            if len(raw) > MAX_RESPONSE_BYTES:
                raise ValueError()
            result = strict_json(raw)
            if not isinstance(result, dict):
                raise ValueError()
            return result
        except (ValueError, UnicodeError, RecursionError) as exc:
            raise DeviceAuthorizationError(
                "INVALID_DEVICE_AUTH_RESPONSE",
                "设备授权服务返回格式无效",
                status_code=502,
            ) from exc

    @staticmethod
    def _raise_response(status: int, raw: bytes):
        code = "DEVICE_AUTH_REJECTED"
        if 300 <= status < 400:
            code = "DEVICE_AUTH_REDIRECT_REJECTED"
        elif len(raw) <= MAX_RESPONSE_BYTES:
            try:
                data = strict_json(raw)
                value = data.get("code") if isinstance(data, dict) else None
                if isinstance(value, str) and re.fullmatch(
                    r"[A-Z][A-Z0-9_]{1,79}", value
                ):
                    code = value
            except (ValueError, UnicodeError, RecursionError):
                pass
        # Do not put response bodies, tokens or private diagnostics into logs/UI.
        raise DeviceAuthorizationError(
            code,
            f"设备授权服务未批准此操作（HTTP {status}）",
            status_code=status,
            transient=status in {429, 502, 503, 504}
            and code not in {"DEVICE_AUTH_NOT_CONFIGURED"},
        )


class DeviceLeaseCache:
    """Non-authoritative signed lease cache; never stores login/access tokens.

    A cache may display a verified last-known device after restart. It cannot grant
    offline admission after process restart because no trusted monotonic anchor
    survives. The original key and account can restore the grant ONLINE, without
    another administrator approval or another device registration.
    """

    def __init__(self, root: Path):
        self.root = Path(root)

    @classmethod
    def for_machine(cls):
        program_data = os.environ.get("ProgramData", "")
        if not program_data or not Path(program_data).is_absolute():
            raise DeviceAuthorizationError(
                "DEVICE_CACHE_UNAVAILABLE", "无法定位设备授权状态目录", status_code=503
            )
        return cls(Path(program_data) / "PublicVideoWorkbench" / "Licensing")

    def path_for(self, user_id: int, thumbprint: str) -> Path:
        if (
            type(user_id) is not int
            or user_id < 1
            or not re.fullmatch(r"[A-Za-z0-9_-]{43}", thumbprint)
        ):
            raise ValueError("invalid cache identity")
        return self.root / thumbprint / f"account-{user_id}.json"

    def save(
        self, user_id: int, thumbprint: str, credentials: VerifiedCredentials
    ) -> None:
        target = self.path_for(user_id, thumbprint)
        if (
            credentials.claims["user_id"] != user_id
            or credentials.claims["cnf"]["jkt"] != thumbprint
        ):
            raise ValueError("cache identity mismatch")
        target.parent.mkdir(parents=True, exist_ok=True)
        data = canonical_json(
            {
                "schema": "publicvideo.device-cache.v1",
                "local_lease": credentials.local_lease,
            }
        )
        temporary = None
        try:
            with tempfile.NamedTemporaryFile(
                mode="w",
                encoding="utf-8",
                dir=target.parent,
                prefix=".lease-",
                suffix=".tmp",
                delete=False,
            ) as output:
                temporary = Path(output.name)
                output.write(data)
                output.flush()
                os.fsync(output.fileno())
            os.replace(temporary, target)
        finally:
            if temporary is not None:
                temporary.unlink(missing_ok=True)

    def hint(
        self, trust: TrustedIssuer, *, user_id: int, thumbprint: str, now: float
    ) -> dict | None:
        try:
            path = self.path_for(user_id, thumbprint)
            with path.open("rb") as source:
                raw = source.read(MAX_RESPONSE_BYTES + 1)
            if len(raw) > MAX_RESPONSE_BYTES:
                return None
            record = strict_json(raw)
            if (
                set(record) != {"schema", "local_lease"}
                or record["schema"] != "publicvideo.device-cache.v1"
            ):
                return None
            claims = trust.verify(
                record["local_lease"],
                typ=LEASE_TYPE,
                user_id=user_id,
                thumbprint=thumbprint,
                now=now,
            )
            return {
                "device_id": claims["device_id"],
                "grant_id": claims["grant_id"],
                "state": "AUTH_REFRESH_REQUIRED",
            }
        except (
            OSError,
            ValueError,
            TypeError,
            KeyError,
            RecursionError,
            DeviceAuthorizationError,
        ):
            return None


class DeviceAuthorizationSession:
    """One logged-in account on THIS backend's machine, not a browser's machine."""

    def __init__(
        self,
        *,
        user_id: int,
        login_token: str,
        trust: TrustedIssuer,
        identity: WindowsDeviceIdentity,
        transport=None,
        cache: DeviceLeaseCache | None = None,
        wall_clock=time.time,
        monotonic_clock=time.monotonic,
    ):
        if type(user_id) is not int or user_id < 1 or not login_token:
            raise ValueError("authenticated account required")
        self.user_id, self.trust, self.identity = user_id, trust, identity
        self._login_token = login_token
        self._transport = (
            transport if transport is not None else DeviceAuthTransport(trust)
        )
        self._cache, self._wall, self._mono = cache, wall_clock, monotonic_clock
        self._lock = threading.RLock()
        self._key = None
        self._credentials = None
        self._anchor_wall = self._anchor_mono = self._refresh_at = 0.0
        self._nonce = None
        self._closed = False
        self.state, self.error_code = "AUTH_REFRESH_REQUIRED", None
        self.cache_warning = False
        self._local_policy = None
        self._background_retry_at = 0.0

    def background_refresh(self) -> bool:
        """Refresh an existing account session, never register or run queued work.

        Skip busy sessions instead of holding up another account. The foreground
        refresh and this method share the same lock, clock anchors and deadlines.
        """
        if not self._lock.acquire(blocking=False):
            return False
        try:
            if self._closed or self.state in {
                "UNREGISTERED",
                "LOGIN_REQUIRED",
                "CLIENT_UPGRADE_REQUIRED",
            }:
                return False
            now = self._mono()
            if self._still_valid() and now < self._refresh_at:
                return False
            if now < self._background_retry_at:
                return False
            # A programming/transport failure must not make a tight retry loop.
            self._background_retry_at = now + 15
            try:
                self.refresh()
                self._background_retry_at = 0.0
            except (DeviceAuthorizationError, DeviceIdentityError):
                # refresh already records the precise state and preserves only a
                # still-valid lease on transient network errors, never its expiry.
                delay = (
                    15
                    if self.state
                    in {"PENDING", "OFFLINE_GRACE", "AUTH_REFRESH_REQUIRED"}
                    else 300
                )
                self._background_retry_at = self._mono() + delay
            return True
        finally:
            self._lock.release()

    def local_policy_mode(self, *, force: bool = False) -> str:
        from .device_local_policy import LocalPolicySession

        with self._lock:
            if self._closed:
                raise DeviceAuthorizationError(
                    "LOGIN_REQUIRED", "请重新登录账号", status_code=401
                )
            if self._local_policy is None:
                self._local_policy = LocalPolicySession(
                    trust=self.trust,
                    transport=self._transport,
                    user_id=self.user_id,
                    account_token=self._login_token,
                    wall_clock=self._wall,
                    monotonic_clock=self._mono,
                )
            return self._local_policy.mode(force=force)

    def _ensure_key(self):
        if self._closed:
            raise DeviceAuthorizationError(
                "LOGIN_REQUIRED", "请重新登录账号", status_code=401
            )
        if self._key is None:
            self._key = self.identity.open_existing()
        if self._key is None:
            raise DeviceAuthorizationError(
                "DEVICE_UNREGISTERED", "此电脑尚未申请设备授权"
            )
        return self._key

    def _proof_request(
        self, purpose: str, *, token: str, bound: bool, payload: dict | None = None
    ) -> dict:
        key = self._ensure_key()
        authorization = ("DPoP " if bound else "Bearer ") + token
        challenge = self._transport.request(
            method="POST",
            path=BASE + "/challenge",
            headers={"Authorization": authorization},
            payload={"public_jwk": key.public_jwk, "purpose": purpose},
        )
        proof = make_proof(
            key,
            self.trust,
            method="GET" if purpose == "status" else "POST",
            path=BASE + "/" + purpose,
            access_token=token,
            nonce=challenge.get("nonce"),
            now=int(self._wall()),
        )
        return self._transport.request(
            method="GET" if purpose == "status" else "POST",
            path=BASE + "/" + purpose,
            headers={"Authorization": authorization, "DPoP": proof},
            payload=payload,
        )

    def _still_valid(self) -> bool:
        if self._credentials is None:
            return False
        elapsed = self._mono() - self._anchor_mono
        wall_elapsed = self._wall() - self._anchor_wall
        if elapsed < 0 or abs(elapsed - wall_elapsed) > 5:
            return False
        return (
            elapsed < self._credentials.claims["exp"] - self._anchor_wall
            and self._wall() < self._credentials.claims["exp"]
        )

    def _failed(self, error: DeviceAuthorizationError | DeviceIdentityError) -> None:
        self.error_code = error.code
        if (
            isinstance(error, DeviceAuthorizationError)
            and error.transient
            and self._still_valid()
        ):
            self.state = "OFFLINE_GRACE"
            # Backoff does not change the lease expiration or monotonic anchor.
            self._refresh_at = self._mono() + 15
            return
        self._credentials, self._nonce = None, None
        states = {
            "DEVICE_UNREGISTERED": "UNREGISTERED",
            "DEVICE_PENDING": "PENDING",
            "DEVICE_REJECTED": "REJECTED",
            "DEVICE_SUSPENDED": "SUSPENDED",
            "DEVICE_REVOKED": "REVOKED",
            "DEVICE_GRANT_EXPIRED": "EXPIRED",
            "LOGIN_REQUIRED": "LOGIN_REQUIRED",
            "CLIENT_UPGRADE_REQUIRED": "CLIENT_UPGRADE_REQUIRED",
        }
        self.state = (
            (
                "KEY_INITIALIZING"
                if error.code == "KEY_SETUP_IN_PROGRESS"
                else "KEY_UNAVAILABLE"
            )
            if isinstance(error, DeviceIdentityError)
            else states.get(error.code, "AUTH_REFRESH_REQUIRED")
        )

    def refresh(self, *, force: bool = False) -> dict:
        with self._lock:
            if not force and self._still_valid() and self._mono() < self._refresh_at:
                return self.summary()
            try:
                key = self._ensure_key()
                previous = self._credentials if self._still_valid() else None
                result = self._proof_request(
                    "refresh" if previous else "exchange",
                    token=previous.access_token if previous else self._login_token,
                    bound=previous is not None,
                )
                # Verify after the round trip so a slow response cannot extend the lease.
                credentials = VerifiedCredentials.from_response(
                    result,
                    self.trust,
                    user_id=self.user_id,
                    thumbprint=key.thumbprint,
                    now=self._wall(),
                )
                self._credentials = credentials
                self._anchor_wall, self._anchor_mono = self._wall(), self._mono()
                self._refresh_at = self._anchor_mono + credentials.refresh_after_seconds
                self._nonce = None
                self.state, self.error_code = "ACTIVE", None
                if self._cache is not None:
                    try:
                        self._cache.save(self.user_id, key.thumbprint, credentials)
                        self.cache_warning = False
                    except OSError:
                        # Cache is optional; do not discard valid online authorization.
                        self.cache_warning = True
                return self.summary()
            except (DeviceAuthorizationError, DeviceIdentityError) as exc:
                self._failed(exc)
                raise

    def register(
        self,
        *,
        label: str = "",
        client_version: str = "",
        operator_sid: str | None = None,
        software_approved: bool = False,
    ) -> dict:
        """Explicit activation action only, never invoked by refresh/cache recovery."""
        with self._lock:
            try:
                if self._closed:
                    raise DeviceAuthorizationError(
                        "LOGIN_REQUIRED", "请重新登录账号", status_code=401
                    )
                if self._key is None:
                    self._key = self.identity.initialize_for_activation(
                        operator_sid=operator_sid, software_approved=software_approved
                    )
                result = self._proof_request(
                    "register",
                    token=self._login_token,
                    bound=False,
                    payload={
                        "protection": self._key.protection,
                        "label": label,
                        "client_version": client_version,
                    },
                )
                # Status is informational. Only signed credentials confer permission.
                self._credentials, self._nonce = None, None
                self.state = self._registration_state(result)
                self.error_code = None
                return result
            except (DeviceAuthorizationError, DeviceIdentityError) as exc:
                self._failed(exc)
                raise

    def software_initialization_permit(self, *, context):
        """Explicit bootstrap consent only; never called by status/refresh/queues."""
        from .device_software_initialization import (
            request_software_initialization_permit,
        )

        with self._lock:
            if self._closed:
                raise DeviceAuthorizationError(
                    "LOGIN_REQUIRED", "请重新登录账号", status_code=401
                )
            return request_software_initialization_permit(
                trust=self.trust,
                transport=self._transport,
                user_id=self.user_id,
                account_token=self._login_token,
                context=context,
                wall_clock=self._wall,
                monotonic_clock=self._mono,
            )

    def register_software(self, *, label="", client_version=""):
        """Explicit compatibility action; never a fallback from normal activation."""
        with self._lock:
            if self._closed:
                raise DeviceAuthorizationError(
                    "LOGIN_REQUIRED", "请重新登录账号", status_code=401
                )
            if self._key is None:
                self._key = self.identity.initialize_software_for_activation(
                    self.software_initialization_permit
                )
            return self.register(label=label, client_version=client_version)

    def repair_key_access(self) -> dict:
        """Explicit existing-key ACL repair; never register or recover a new key."""
        with self._lock:
            try:
                if self._closed:
                    raise DeviceAuthorizationError(
                        "LOGIN_REQUIRED", "请重新登录账号", status_code=401
                    )
                self._credentials, self._nonce = None, None
                if self._key is not None:
                    self._key.close()
                    self._key = None
                self.identity.repair_operator_access()
                return self.refresh(force=True)
            except (DeviceAuthorizationError, DeviceIdentityError) as exc:
                self._failed(exc)
                raise

    def _registration_state(self, result: dict) -> str:
        state = result.get("status")
        if not isinstance(state, str) or state not in {
            "UNREGISTERED",
            "PENDING",
            "ACTIVE",
            "REJECTED",
            "SUSPENDED",
            "REVOKED",
            "EXPIRED",
        }:
            raise DeviceAuthorizationError(
                "INVALID_DEVICE_AUTH_RESPONSE", "设备状态响应格式无效", status_code=502
            )
        if (
            result.get("thumbprint") is not None
            and result.get("thumbprint") != self._key.thumbprint
        ):
            raise DeviceAuthorizationError(
                "DEVICE_IDENTITY_MISMATCH", "设备状态与本机密钥不一致"
            )
        return state

    def status(self) -> dict:
        with self._lock:
            try:
                result = self._proof_request(
                    "status", token=self._login_token, bound=False
                )
                state = self._registration_state(result)
                if state != "ACTIVE":
                    self._credentials, self._nonce = None, None
                self.state, self.error_code = state, None
                return result
            except (DeviceAuthorizationError, DeviceIdentityError) as exc:
                self._failed(exc)
                raise

    def require_local(self, scope: str) -> dict:
        if scope not in {"local:draft", "local:render"}:
            raise ValueError("local admission scope required")
        with self._lock:
            try:
                self.refresh()
            except DeviceAuthorizationError as exc:
                if not exc.transient or not self._still_valid():
                    raise
            if not self._still_valid():
                self.state = "AUTH_REFRESH_REQUIRED"
                raise DeviceAuthorizationError(
                    "AUTH_REFRESH_REQUIRED", "请联网重新校验设备授权", status_code=401
                )
            if scope not in self._credentials.claims["scopes"]:
                raise DeviceAuthorizationError(
                    "DEVICE_SCOPE_DENIED", "此设备未获得该功能的授权"
                )
            try:
                assert_key_available(self._ensure_key())
            except (DeviceIdentityError, DeviceAuthorizationError) as exc:
                self._failed(exc)
                raise
            # This is a diagnostic/admission binding, not a permanent queue pass.
            return {
                name: self._credentials.claims[name]
                for name in (
                    "user_id",
                    "device_id",
                    "grant_id",
                    "grant_revision",
                    "policy_revision",
                )
            }

    def request_headers(
        self, *, method: str, path: str, scope: str | None
    ) -> dict[str, str]:
        with self._lock:
            self.trust.request_uri(path)
            self.refresh()
            if not self._still_valid():
                raise DeviceAuthorizationError(
                    "AUTH_REFRESH_REQUIRED", "请重新校验设备授权", status_code=401
                )
            if scope is not None and scope not in self._credentials.claims["scopes"]:
                raise DeviceAuthorizationError(
                    "DEVICE_SCOPE_DENIED", "此设备未获得该功能的授权"
                )
            key, token = self._ensure_key(), self._credentials.access_token
            token_hash = sha256_b64(token)
            if (
                self._nonce is None
                or self._nonce[1] <= self._mono()
                or self._nonce[2] != token_hash
            ):
                result = self._transport.request(
                    method="POST",
                    path=BASE + "/challenge",
                    headers={"Authorization": "DPoP " + token},
                    payload={"public_jwk": key.public_jwk, "purpose": "request"},
                )
                ttl = result.get("expires_in")
                if type(ttl) is not int or ttl < 1:
                    raise DeviceAuthorizationError(
                        "INVALID_DEVICE_CHALLENGE", "设备验证挑战有效期无效"
                    )
                self._nonce = (
                    result.get("nonce"),
                    self._mono() + min(90, ttl - 1),
                    token_hash,
                )
            proof = make_proof(
                key,
                self.trust,
                method=method,
                path=path,
                access_token=token,
                nonce=self._nonce[0],
                now=int(self._wall()),
            )
            return {"Authorization": "DPoP " + token, "DPoP": proof}

    def summary(self) -> dict:
        with self._lock:
            valid = self._still_valid()
            state = (
                self.state
                if self.state not in {"ACTIVE", "OFFLINE_GRACE"} or valid
                else "AUTH_REFRESH_REQUIRED"
            )
            result = {
                "state": state,
                "error_code": self.error_code,
                "cache_warning": self.cache_warning,
                "user_id": self.user_id,
                "thumbprint": self._key.thumbprint if self._key else None,
                "protection_report": self._key.protection if self._key else None,
                "protection_verified": False,
            }
            if self._credentials:
                result.update(
                    {
                        name: self._credentials.claims[name]
                        for name in ("device_id", "grant_id", "exp")
                    }
                )
            return result

    def close(self) -> None:
        with self._lock:
            if self._local_policy is not None:
                self._local_policy.clear()
                self._local_policy = None
            self._credentials, self._nonce, self._login_token = None, None, ""
            self._closed = True
            if self._key is not None:
                self._key.close()
                self._key = None
            self.state = "LOGIN_REQUIRED"
